### PR TITLE
Allow unused_identifier in function definitions in Elixir

### PR DESCRIPTION
**Pull request by hauleth**

### CODE REVIEW

**Review by theHamsta — APPROVED** [review]

File: queries/elixir/highlights.scm [hunk 1]
```diff
@@ -156,6 +156,8 @@
    [(identifier) @parameter
     (tuple (identifier) @parameter)
     (list (identifier) @parameter)
+    (unused_identifier) @comment
+    (map) @comment
```
Comment (theHamsta) [inline queries/elixir/highlights.scm:160]: Why is map a comment?

**Review by theHamsta — CHANGES_REQUESTED** [review]
~~There are some queries without captures~~

**Review by theHamsta — APPROVED** [review]

**Comment by theHamsta** [comment]
@nifoc could you have a look on this?

**Comment by nifoc** [comment]
I think this was removed by accident (from `Definition with (some) arguments and guard(s)`):

```
(_
 (keyword_list (identifier) @parameter))
```

Otherwise LGMT ❤️

**Comment by theHamsta** [comment]
@hauleth could you address the comment of @nifoc ?

**Comment by hauleth** [comment]
@theHamsta yes, soonish

**Comment by hauleth** [comment]
@theHamsta I just removed everything and allowed everything to be called in the parameters, as it is how Elixir works:

```elixir
def foo(bar(a: a)) do
  # …
end
```

Is completely valid code that can be spotted in real projects (when working with Erlang records).

**Comment by nifoc** [comment]
Sadly this also removes all other `@parameter` highlights.

tree-sitter doesn't really support languages with pattern matching (in function heads) at the moment, which is why these queries are so *verbose* in the first place. Some patterns are explicitly listed in the query file to at least highlight the most common `@parameter`s in pattern matches.

Currently I would say it's nearly impossible to correctly highlight every possible pattern match ...
[This discussion](https://github.com/tree-sitter/tree-sitter/discussions/1096) also mentions the above issue and links to some further sources/issues.

**Comment by hauleth** [comment]
@nifoc that discussion is pointless as even with such thing supporting all possible patterns as `quote` call can be used as a pattern. Also I think that *missing* feature is better than having feature that works only partially. That is why I think that just removing support for colouring `@parameter` and instead having full support for colouring function definitions is worth it.

And this is how it looks right now:

Zrzut ekranu 2021-07-13 o 22 27 56

**Comment by nifoc** [comment]
Okay, now I understand.

To add a catch all query that highlights the function name, but doesn't highlight parameters (because none of the more specific queries match), I would suggest adding the following before `; Definition with (some) arguments and (optional) defaults`:

```
(call (function_identifier) @keyword.function
 (call
  function: (function_identifier) @function)
 (#lua-match? @keyword.function "^def%a*$"))
```

This will match both `foo` (and mark them as `@function`):

Bildschirmfoto 2021-07-13 um 22 41 01

**Comment by nifoc** [comment]
Sorry if my previous comment caused any confusion. The solution could cause tree-sitter to run into an internal limit, which results in only some parameters being highlighted for functions with many parameters.

The following should match (and highlight) `foo` in both function definitions:
(Notice the `?` after the `(arguments ...)`)

```
; Definition with (some) arguments and (optional) defaults
(call (function_identifier) @keyword.function
 (call
  function: (function_identifier) @function
  (arguments
   [(identifier) @parameter
    (tuple (identifier) @parameter)
    (list (identifier) @parameter)
    (_
     (keyword_list (identifier) @parameter))
    (binary_op
     left: (identifier) @parameter
     operator: "\\\\")])?)
 (#lua-match? @keyword.function "^def%a*$"))

; Definition with (some) arguments and guard(s)
(call (function_identifier) @keyword.function
 (binary_op
  left:
   (call
    function: (function_identifier) @function
    (arguments
     [(identifier) @parameter
      (tuple (identifier) @parameter)
      (list (identifier) @parameter)
      (_
       (keyword_list (identifier) @parameter))])?)
  operator: "when")
 (#lua-match? @keyword.function "^def%a*$"))
```

Looks like this in the playground:

Bildschirmfoto 2021-07-13 um 23 01 23

**Comment by hauleth** [comment]
However it will again not apply the `@parameter` to the passed arguments, yes? That was the core of the problem.

It seems that if you try to do almost any pattern matching in the function head, then it will fail. It seems that even the really basic pattern matching will make the parser to fail.

Zrzut ekranu 2021-07-14 o 14 57 41

**Comment by nifoc** [comment]
> However it will again not apply the `@parameter` to the passed arguments, yes?

In most of your examples it will not highlight the `@parameter`.
(But with the `?` added it will at least highlight the `@function` in all cases)

Bildschirmfoto 2021-07-14 um 15 22 23

> That was the core of the problem.

Just so we're definitely talking about the same thing, can you maybe spell out what you wanted to achieve with this PR?
I was under the impression that it's mainly about the missing `@function` highlight.

**Comment by theHamsta** [comment]
@nifoc I'm really not an elixir expert. Can you approve this as soon as the PR is ready? I have the feeling I'm not really helpful regarding elixir...

**Comment by nifoc** [comment]
@theHamsta I think the bot merged this by accident?

(It doesn't break anything, just removes some `@parameter` highlights)

**Comment by stsewd** [comment]
> @theHamsta I think the bot merged this by accident?
> 
> (It doesn't break anything, just removes some `@parameter` highlights)

it was by accident, but let's just keep adding improvements in another PR, as you said this doesn't break anything

**Pull request merged**